app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,168 @@
+"""
+Test for recipe API methods.
+"""
+
+from django.test import TestCase
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from core.models import (Recipe, Ingredient)
+
+RECIPES_URL = reverse('recipe:recipe-list')
+
+
+def detail_url(recipe_id):
+    """Create and return a recipe detail URL."""
+    return reverse('recipe:recipe-detail', args=[recipe_id])
+
+
+def create_recipe(name="Sample Recipe", description="Sample Recipe Description."):
+    """Create and return a sample recipe object."""
+    return Recipe.objects.create(name=name, description=description)
+
+
+def create_ingredient(recipe, name="Sample Ingredient Name"):
+    """Create and return a new ingredient object."""
+    return Ingredient.objects.create(name=name,
+                                     recipe=recipe)
+
+
+class PublicRecipeApiTests(TestCase):
+    """Unauthenticated Recipe API tests"""
+
+    def setUp(self):
```

Comment:
I suggest yo to use `setUpTestData` instead.

`setUpTestData` is called once at the beginning of the test run for class-level setup. While `setUp` is called before every test function.

Reply:
Thanks! But when I tried this, I found an issue. Seems like setUpTestData is a static class method and i have to use it like this:

```
 @classmethod
    def setUpTestData(cls):
        cls.client = APIClient()
```

Which means that in all the test cases, instead of accessing `self.client`, I have to use the class variable `PublicRecipeApiTests.client` to make them work. 

Is there some better way to do this? 🤔  Since in this setup, I am only defining the APIClient from rest framework that I need in all the test cases.

Reply:
That's odd. I think you should be able to acess it by self.client :thinking:

Reply:
Actually, I was able to access `self.client` but the problem was something else with APIClient. Due to some wierd behaviour which I couldn't figure out properly, the line `self.client.post(path, payload, format='json')` does not work if i define the client as a class variable and access it by self.client but it did work if I use CLASNAME.client or if I define the client as instance variable in setUp. 

For the propblematic scenario when I use `self.client`, the root cause is something related to formatters and encoders defined inside APIClient implementation and even if I used the format='json', it wouldn't recognise that and encode the request as multipart/form. 

But a workaround was to provide the content type and encode the payload manually like:

```
self.client.patch(url, json.dumps(payload), content_type='application/json')
``` 

and this works. I updated the PR to have this change. Although I still didn't get fully why the APIClient post and patch methods behave differently on the client being defined as a class variable or an instance variable.